### DISCUSSION TITLE
obtain content checksum (without metadata)

### DISCUSSION THREAD

**Discussion by eadmaster** [qpdf/qpdf#1220] [Q&A]
is it possible to obtain a checksum of a pdf file excluding its metadata?

use case: compare 2 pdf files by content

related:
- https://github.com/qpdf/qpdf/issues/447
- https://stackoverflow.com/questions/28826310/md5-checksum-of-pdf-file
- https://github.com/mozilla/pdf.js/issues/10569

If not supported, do you know other tools that can do that?

**Comment by jberkenbilt** [comment]
I'm not sure exactly what you mean by this. Compare two pdf files by content sounds like a tall order since there are so many ways of rearranging the bytes in a PDF file without meaningfully changing its contents. I have a little Python script that uses tiff tools and ghostscript to compare PDFs as images. This works pretty well to see if two files render the same. With qpdf, you could run both files through qpdf with --deterministic-id and compare them, but it would not be perfect. If they matched, you would have a guarantee, but you will get many false negatives. I'm attaching my Python script in case it is useful.

**Comment by jberkenbilt** [comment]
```python
#!/usr/bin/env python3
import os
import subprocess
import sys
import tempfile

whoami = os.path.basename(sys.argv[0])


class Main:
    def usage(self):
        exit(f'''
Usage: {whoami} a.pdf b.pdf
Compare a.pdf and b.pdf rendered as images
''')

    def main(self, args=sys.argv[1:], prog=whoami):
        if len(args) != 2:
            self.usage()
        file1, file2 = args
        with tempfile.TemporaryDirectory() as tmpdir:
            self.compare_pdfs(tmpdir, file1, file2)

    def convert(self, pdf, tiff):
        subprocess.run(['gs', '-q', '-r200', '-dNOPAUSE', '-sDEVICE=tiff24nc',
                        '-dBATCH', f'-sOutputFile={tiff}', pdf],
                       stdout=subprocess.DEVNULL,
                       stderr=subprocess.DEVNULL,
                       check=True)

    def compare_pdfs(self, tmpdir, file1, file2):

        # We discard gs's stderr since it has sometimes been known to
        # complain about files that are not bad. In particular, gs 9.04
        # can't handle empty xref sections such as those found in the
        # hybrid xref cases. We don't really care whether gs complains or
        # not as long as it creates correct images. If it doesn't create
        # correct images, the test will fail, and we can run manually to
        # see the error message. If it does, then we don't care about the
        # warning.

        f1 = os.path.join(tmpdir, 'a.tif')
        f2 = os.path.join(tmpdir, 'b.tif')
        self.convert(file1, f1)
        self.convert(file2, f2)
        p = subprocess.run(['tiffcmp', '-t', f1, f2])
        if p.returncode == 0:
            print(f'{whoami}: {file1} and {file2} render identically')
        else:
            exit(f'{whoami}: {file1} and {file2} render differently')


if __name__ == '__main__':
    try:
        Main().main()
    except KeyboardInterrupt:
        exit(130)
```

**Comment by eadmaster** [comment]
"by content" i mean that the files must have the exact same content bit-wise, so comparing the rendered images won't do.

The use case is to find pdf duplicates with different filenames or edited metadata.

This is the same as comparing 2 media files with hashes as supported by the [ffmpeg `streamhash` filter](https://stackoverflow.com/questions/63608312/how-to-md5-the-video-track-only-in-ffmpeg).

**Comment by jberkenbilt** [comment]
qpdf doesn't have anything like this out of the box. With a precise definition of what the metadata is, I suspect something like this could be written, but I'm not sure exactly which parts of a PDF would be considered metadata for purposes of this comparison. The PDF file structure allows much more variation than image and video formats, and editing metadata would surely require rewriting the file, which would change offsets and likely other things not related to the content, making a bitwise comparison not match. In the absence of knowledge of any particular specification for what you're looking for, my impulse would be to write something that would read two files, remove whatever parts are considered metadata (probably info dictionary keys and the /Metadata object from the root dictionary), then write them out with deterministic /ID and compare the results.

**Reply by eadmaster** [reply]
Another approach i'm thinking is to treat pdf stream objects as ffmpeg "tracks", and hash them separately. So for example with [this pdf](https://segaretro.org/File:BeepMD_JP_1994-04.pdf) we would have a separate hash for each image:
````
$ qpdf  --show-pages  --with-images --show-crc32  BeepMD_JP_1994-04.pdf
page 1: 3 0 R
  images:
    /I0_0: 2 0 R, 1233 x 1520  470dac87
  content:
    4 0 R
page 2: 9 0 R
  images:
    /I1_0: 8 0 R, 1233 x 1520  97527dbc
  content:
    10 0 R
page 3: 15 0 R
  images:
    /I2_0: 14 0 R, 1233 x 1520  4f9f5812
  content:
    16 0 R
````

I prefer CRC32 hashes since these are usually stored inside compressed archives and would make the comparison quicker.

**Reply by jberkenbilt** [reply]
If you took that approach, using qpdf JSON with `--json-stream-data=file`, could be helpful as you would have a separate file for each stream's contents. You could also use `--json-stream-data=inline` (the default) in which case each stream's data is a base64-encoded string. Of course, you can also do that with straight qpdf.
